src/expr/node_algorithm.cpp

```diff
@@ -273,5 +273,121 @@ void getSymbols(TNode n,
   } while (!visit.empty());
 }
 
+Node substituteCaptureAvoiding(TNode n, Node src, Node dest)
+{
+  if (n == src)
+  {
+    return dest;
+  }
+  if (src == dest)
+  {
+    return n;
+  }
+  std::vector<Node> srcs;
+  std::vector<Node> dests;
+  srcs.push_back(src);
+  dests.push_back(dest);
+  return substituteCaptureAvoiding(n, srcs, dests);
+}
+
+Node substituteCaptureAvoiding(TNode n,
+                               std::vector<Node>& src,
+                               std::vector<Node>& dest)
+{
+  std::unordered_map<TNode, Node, TNodeHashFunction> visited;
+  std::unordered_map<TNode, Node, TNodeHashFunction>::iterator it;
+  std::vector<TNode> visit;
+  TNode curr;
+  visit.push_back(n);
+  Assert(src.size() == dest.size(),
+         "Substitution domain and range must be equal size");
+  do
+  {
+    curr = visit.back();
+    visit.pop_back();
+    it = visited.find(curr);
+
+    if (it == visited.end())
+    {
+      auto itt = std::find(src.begin(), src.end(), curr);
+      if (itt != src.end())
+      {
+        Assert(std::distance(src.begin(), itt) >= 0
+               && static_cast<unsigned>(std::distance(src.begin(), itt))
+                      < dest.size());
+        Node n = dest[std::distance(src.begin(), itt)];
+        visited[curr] = n;
+        continue;
+      }
+      if (curr.getNumChildren() == 0)
+      {
+        visited[curr] = curr;
+        continue;
+      }
+
+      visited[curr] = Node::null();
+      // if binder, rename variables to avoid capture
+      if (curr.isClosure())
+      {
+        std::vector<Node> vars;
+        std::vector<Node> renames;
+
+        NodeManager* nm = NodeManager::currentNM();
+        for (const Node& v : curr[0])
+        {
+          vars.push_back(v);
+          renames.push_back(nm->mkBoundVar(v.getType()));
+        }
+        // have new vars -> renames subs in the beginning of current sub
+        src.insert(src.begin(), vars.begin(), vars.end());
+        dest.insert(dest.begin(), renames.begin(), renames.end());
```

Comment:
This is not critical for this PR but I think that it would be a bit more efficient to insert the new variables at the end of `src`/`dest` (and without creating `vars` and `renames`) and then use `std::find(src.rbegin(), src.rend(), curr)` on line 312.

Reply:
Good suggestion! I changed accordingly

Reply:
Cool, this looks good :)